### ISSUE TITLE
execute() with empty list should raise a deprecation warning, both core / ORM

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#9647]
### Discussed in https://github.com/sqlalchemy/sqlalchemy/discussions/9645

<div type='discussions-op-text'>

<sup>Originally posted by **liujin1993** April 13, 2023</sup>
#### question
sqlalchemy version: 2.0.9
python version: 3.10.6

Hi, when i use `session.execute(insert(Foo), data_list)` replacing `bulk_save_objects()` to bulk insert rows, I encounter a strange scenario as follow:
```python
engine = create_engine(
    'mysql+pymysql://...', 
    pool_recycle=3600, echo=True
)
Session = sessionmaker(bind=engine)
with Session() as session:
    session.bulk_save_objects([])      #  there is no generated sql
    session.execute(insert(Foo), [])   #  sqlalchemy generate a null sql: INFO sqlalchemy.engine.Engine INSERT INTO foo () VALUES ()
    session.commit()
```
![image](https://user-images.githubusercontent.com/31949674/231700610-9b398e64-406d-4309-b751-cf81c54afc37.png)

Sqlalchemy insert a null row when value is an empty list. Is it expected?

#### reference
https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements

</div>

Mike on https://github.com/sqlalchemy/sqlalchemy/discussions/9645#discussioncomment-5604432 suggested:
> his is a bug and this should raise well before it gets to insertmanyvalues. IMO connection.execute(..., []) of any kind should raise. While it's true that dbapi.executemany() will usually raise for this, at this point we have too much going on to let it get that far. I would say the best place is in the DefaultExecutionContext where we decide that we are an "executemany". then "if not params: raise". that's it. I think only for exec_driver_sql(), if even that, would we leave alone, as though some driver had a way to accommodate an empty collection in executemany(). for other statement execution styles that have an executemany it would raise.

**Comment by CarlosEduR** [comment]
@CaselIT is PR with tests welcome on this issue?

**Comment by CaselIT** [comment]
Sure, mike provided some guidance on the implementation, feel free to ask if you have additional questions!

**Comment by CarlosEduR** [comment]
Appreciate it!

I've started the implementation, will be opening the PR soon, just had one question while running the tests. 
Based on what Mike said: `connection.execute(..., []) of any kind should raise.`, I would like to confirm about this case, one test that is failing is this one:
```
def test_empty_insert(self, connection):
    """test that execute() interprets [] as a list with no params"""
    users_autoinc = self.tables.users_autoinc

    connection.execute(
        users_autoinc.insert().values(user_name=bindparam("name", None)),
        [],
    )
    eq_(connection.execute(users_autoinc.select()).fetchall(), [(1, None)])
```

This

**Comment by CaselIT** [comment]
At the moment the `[]` is ignored and the bindparam is used instead, right?

I'll leave the reply to mike here

**Comment by CarlosEduR** [comment]
> Member

Yes, here is the code I am running and the logged queries:
```
with Session() as session:
    session.execute(insert(Foo), [])
    session.execute(insert(Foo).values(name=bindparam("user_nane", None)), [])
    session.commit()
```
```
2023-10-02 15:09:15,953 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-10-02 15:09:15,953 INFO sqlalchemy.engine.Engine INSERT INTO foo () VALUES ()
2023-10-02 15:09:15,953 INFO sqlalchemy.engine.Engine [generated in 0.00008s] {}
2023-10-02 15:09:15,954 INFO sqlalchemy.engine.Engine INSERT INTO foo (name) VALUES (%(user_nane)s)
2023-10-02 15:09:15,954 INFO sqlalchemy.engine.Engine [generated in 0.00008s] {'user_nane': None}
2023-10-02 15:09:15,955 INFO sqlalchemy.engine.Engine COMMIT
```

**Comment by CaselIT** [comment]
@zzzeek do you think that the above case raise or not? I kinda see reason to having any of the two options

**Comment by zzzeek** [comment]
the problem here is that if you say connection.execute(table.insert(), None), it's going to insert an empty row with default values. that's correct.    If you say connection.execute(table.insert(), []), that's more ambiguous (in 1.4 it was just an empty parameter set) since that argument is only supposed to be a dictionary or list of dictionaries in 2.0.   so if you send an empty list, that means you dont actually want to run an INSERT.  should raise, however for connection, I dont think we could change that in 2.0, would have to be 2.1 if it's truly treating the empty list as a single row with no params.

for ORM, it should raise for an empty list, but not for an empty dict (and empty tuple is again kind of ambiguous now)

**Comment by zzzeek** [comment]
this is the change for 2.1:

```diff
diff --git a/lib/sqlalchemy/engine/_py_util.py b/lib/sqlalchemy/engine/_py_util.py
index 538c075a2b..1041a3dda7 100644
--- a/lib/sqlalchemy/engine/_py_util.py
+++ b/lib/sqlalchemy/engine/_py_util.py
@@ -30,6 +30,11 @@ def _distill_params_20(
             raise exc.ArgumentError(
                 "List argument must consist only of tuples or dictionaries"
             )
+        elif not params:
+            raise exc.ArgumentError(
+                "Empty list of parameter dictionaries passed; "
+                "no statement to execute"
+            )
 
         return params
     elif isinstance(params, dict) or isinstance(

```

we can bring this through a whole deprecation process if we want.   similar behavior would need to occur for ORM and this is likely separate

**Comment by sqla-tester** [comment]
**Carlos Sousa** has proposed a fix for this issue in the **main** branch:

**Raise exception for execute() when parameter is an empty list** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5101

**Comment by CaselIT** [comment]
This was downgraded to a warning in 2.1 since it looks like too disruptive of a change to do without warning first.

**Comment by JacobHayes** [comment]
For whatever reason, this doesn't raise a warning with `sa.insert` (the construct used in the OP) nor dialect specific `insert`s, only the `Table.insert` construct - should this be reopened or a new issue created?

Separately, could this (fixed) warning be backported to 2.0, if 2.1 is still a bit away? I'm not certain whether you all allow new warnings to be added within a "MAJOR.MINOR", but if it stayed a warning in 2.1 too maybe that's ok?

---

Repro:

If you run this script from `main` or the commit that closed this issue, a warning is raised for `MyTable.__table__.insert()`, but not from the other constructs. And on 2.0 it "of course" just inserts an empty row without any flag. 😅


```python
import sqlalchemy as sa
import warnings
from sqlalchemy.dialects import sqlite
from sqlalchemy.orm import DeclarativeBase, Mapped, Session, mapped_column

warnings.simplefilter("always")  # Don't cache warning messages


class Base(DeclarativeBase):
    pass


class MyTable(Base):
    __tablename__ = "my_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    # Removing the default causes the insert to error on null values
    b: Mapped[int] = mapped_column(default=0)


engine = sa.create_engine("sqlite:///:memory:")
Base.metadata.create_all(engine)
with Session(engine) as session:
    session.execute(MyTable.__table__.insert(), [])  # Raises a warning *on main*
    session.execute(sa.insert(MyTable), [])  # Doesn't warn at all
    session.execute(sqlite.insert(MyTable), [])  # Doesn't warn at all

print("Done...")
```

I was testing across versions with `uv`:
```sh
uvx --python 3.12 --with 'sqlalchemy @ git+https://github.com/sqlalchemy/sqlalchemy.git@af655e55d3cb15895901de7803726ee434389445' python test.py
uvx --python 3.12 --with 'sqlalchemy @ git+https://github.com/sqlalchemy/sqlalchemy.git@main' python test.py
uvx --python 3.12 --with 'sqlalchemy @ git+https://github.com/sqlalchemy/sqlalchemy.git@rel_2_0_38' python test.py
```
